### PR TITLE
Guard against tokenizing strings > 20K

### LINKED ISSUE #556: Calva locking up when opening files with very long lines
When opening a file with very long lines all Calva commands stop working.

A symptom is that even just hitting backspace stops doing what it should. Another is that the CPU gets very a lot of load.

I think this is because that the tokenizer (probably the `TokenCurosr`, but it could also be the `Calva highlight` tokenizer) locks up trying to split the line up with its regular expressions.

A ”solution” is to do what the VS Code built-in syntax highlighter does: Recognise that tokenisation will take a long time and just skip doing it then.

### PR DESCRIPTION
**Pull request by PEZ**
Work in progress...

## What has Changed?
I've put a guard in `lexer.ts` to guard against tokenization of long lines. Right now using 20K length hardcoded, but probably should be using vscode's tokenization maxlength setting.

The current approach is to recognize when a line is too long and only represent it as a `too-long-line` token.

Fixes #556

## My Calva PR Checklist
I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Made sure I am directing this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I am changed the default PR base branch, so that it is not `master`. (Sorry for the nagging.)
- [x] Tested the VSIX built from the PR (well, if this is a PR that changes the source code.) You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test. (For now you'll need to opt in to the CircleCI _New Experience_ UI to see the Artifacts tab, because bug.)
     - [x] Tested the particular change
     - [x] Figured if the change might have some side effects and tested those as well.
     - [x] Smoke tested the extension as such.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
     - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.

## The Calva Team PR Checklist:

Before merging we (at least one of us) have:

- [ ] Made sure the PR is directed at the `dev` branch (unless reasons).
- [ ] Read the source changes.
- [ ] Given feedback and guidance on source changes, if needed. (Please consider noting extra nice stuff as well.)
- [ ] Tested the VSIX built from the PR (well, if this is a PR that changes the source code.)
     - [ ] Tested the particular change
     - [ ] Figured if the change might have some side effects and tested those as well.
     - [ ] Smoke tested the extension as such.
- [ ] If need be, had a chat within the team about particular changes.

Ping @pez, @kstehn, @cfehse, @bpringe

### CODE REVIEW

**Comment by bpringe** [comment]
Sounds like a good approach. I agree that using VS Code's setting would probably be ideal.

**Comment by PEZ** [comment]
Now I fixed a proper scanner object for tokenising the too-long-strings.

I can't figure out how to get hold of the VSCode setting for tokenisation max-length though...

**Comment by PEZ** [comment]
Now added an `initScanner` function that needs to get called before lexing anything. It's a bit messy, but I think it is the right place to pass in the vscode max tokenisation setting.

File: src/webview.ts [hunk 1]
```diff
@@ -4,7 +4,10 @@ var Ansi = require('ansi-to-html');
 import "../assets/styles/webview.scss";
 import escapeHTML = require("escape-html");
 import * as paredit from "./cursor-doc/paredit";
-import { ModelEdit, ModelEditSelection } from "./cursor-doc/model";
+import { ModelEdit, ModelEditSelection, initScanner } from "./cursor-doc/model";
+
+const MAX_LINE_TOKENIZATION_LENGTH = 20000;
```
Comment (bpringe) [inline src/webview.ts]: Can we also use the vs code setting here? And do we want to?

Comment (PEZ) [reply src/webview.ts]: It would be nice to do that, but it wasn't immediately obvious to me how to do it.

**Comment by PEZ** [comment]
We will have to make it prettier later. This tokenization freeze is causing too much troubles out there.

**Pull request merged**